### PR TITLE
Fix issue #12883

### PR DESCRIPTION
**Pull request by etruong42**
Fixed https://github.com/hashicorp/terraform-provider-google/issues/12883

Make one CloudResourceManager call per monitored_project terraform resource rather than per (monitored_project * "sibling" projects)

If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [ ] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/develop/run-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
monitoring: fixed scaling issues when deploying terraform changes with many `google_monitoring_monitored_project`
```

### CODE REVIEW

**Review by hao-nan-li — COMMENTED** [review]
Looks like there are some build errors. 
https://googlecloudplatform.github.io/magic-modules/develop/run-tests/ might help.

**Review by hao-nan-li — COMMENTED** [review]
Looks like the tests failed due to     testing_new.go:84: Error running post-test destroy, there may be dangling resources: MonitoringMonitoredProject still exists at https://monitoring.googleapis.com/v1/locations/global/metricsScopes/ci-test-project-188019. Do you have similar behaviour when you run the tests locally?

**Comment by etruong42** [comment]
No - they all return ok.

**Comment by hao-nan-li** [comment]
> No - they all return ok.

Could you upload the logs?

**Comment by etruong42** [comment]
https://pastebin.com/BTpJymfe

**Comment by hao-nan-li** [comment]
> https://pastebin.com/BTpJymfe

The log shows that build tests passed. 

Here, the failing tests are acceptance tests or automated tests. You can find instructions to run them at https://googlecloudplatform.github.io/magic-modules/develop/run-tests/#run-automated-tests.

**Review by davidbat — COMMENTED** [review]
`if strings.HasSuffix(monitoredProjectName.(string), fmt.Sprintf("/%s", expectedProjectNumber)) {`

This is technically unsafe with a super rare edge case where a project number could be a substring of another
For example if you're looking for project 12345678, it would still match
/location/global/metricsScopes/987654321/project/123456789

I would instead check for the string regex fmt.Sprintf("projects/%s$", expectedProjectNumber))

**Comment by etruong42** [comment]
> > https://pastebin.com/BTpJymfe
> 
> The log shows that build tests passed.
> 
> Here, the failing tests are acceptance tests or automated tests. You can find instructions to run them at https://googlecloudplatform.github.io/magic-modules/develop/run-tests/#run-automated-tests.

Those fail right out of the gate in just setting up a project:

```
$ make testacc TEST=./google/services/monitoring TESTARGS='-run=TestAccMonitoringMonitoredProject_monitoringMonitoredProjectBasicExample'
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google/services/monitoring -v -run=TestAccMonitoringMonitoredProject_monitoringMonitoredProjectBasicExample -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccMonitoringMonitoredProject_monitoringMonitoredProjectBasicExample
=== PAUSE TestAccMonitoringMonitoredProject_monitoringMonitoredProjectBasicExample
=== CONT  TestAccMonitoringMonitoredProject_monitoringMonitoredProjectBasicExample
    vcr_utils.go:152: Step 1/2 error: Error running apply: exit status 1

        Error: error creating project tf-test-m-idi8gfqy8ti0 (tf-test-m-idi8gfqy8ti0-display): googleapi: Error 400: field [parent] has issue [Parent id must be numeric.]
        Details:
        [
          {
            "@type": "type.googleapis.com/google.rpc.BadRequest",
            "fieldViolations": [
              {
                "description": "Parent id must be numeric.",
                "field": "parent"
              }
            ]
          },
          {
            "@type": "type.googleapis.com/google.rpc.Help",
            "links": [
              {
                "url": "https://cloud.google.com/resource-manager/reference/rest/v1/projects"
              }
            ]
          }
        ]
        , badRequest. If you received a 403 error, make sure you have the `roles/resourcemanager.projectCreator` permission

          with google_project.basic,
          on terraform_plugin_test.tf line 7, in resource "google_project" "basic":
           7: resource "google_project" "basic" {

--- FAIL: TestAccMonitoringMonitoredProject_monitoringMonitoredProjectBasicExample (3.30s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-google/google/services/monitoring	4.029s
FAIL
make: *** [testacc] Error 1
```

**Comment by hao-nan-li** [comment]
> 

Looks like you don't have the permission to create a new project. You'd need to grant yourself this permission on pantheon...

**Comment by etruong42** [comment]
I got a 400 error and not a 403 error? I think the message says "if you received a 403 error", in which case, the message seems to indicate that I got a 400 error instead?

File: mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb [hunk 1]
```diff
@@ -12,23 +12,33 @@
...
+monitoredProjects, _ = res["monitoredProjects"].([]interface{})
 
-name := res["name"].(string)
-name = tpgresource.GetResourceNameFromSelfLink(name)
+// Convert configured terraform monitored_project resource name to a projectId
+expectedProject, configProjectErr := config.NewResourceManagerClient(config.UserAgent).Projects.Get(d.Get("name")).Do()
+if configProjectErr != nil {
+	return nil, configProjectErr
+}
+expectedProjectId = expectedProject.ProjectId
```
[the remark this answers is not in the document]
Comment (etruong42) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb]: Done.

File: mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb [hunk 2]
```diff
@@ -12,23 +12,38 @@
...
-	res["name"] = name
-} else if name != "" {
-	project, err := config.NewResourceManagerClient(config.UserAgent).Projects.Get(name).Do()
-	if err != nil {
-		return nil, err
+	monitoredProjectName := monitoredProject["name"]
+	// MonitoredProject names have the format locations/global/metricsScopes/[metricScopeProjectNumber]/projects/[projectNumber]
+	// `res` contains the MonitoredProjects of the relevant metrics scope
+	log.Printf("[DEBUG] Matching ProjectNumbers: %s to %s.", expectedProjectNumber, monitoredProjectName)
+	if strings.HasSuffix(monitoredProjectName.(string), fmt.Sprintf("/%s", expectedProjectNumber)) {
```
[the remark this answers is not in the document]
Comment (etruong42) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:42]: I'm using "HasSuffix" - I would not expect `/location/global/metricsScopes/987654321/project/123456789` to have the _suffix_ `/12345678`. I would expect this bug if it were some sort of `Contains` or `HasSubstr` method.

File: mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb [hunk 3]
```diff
@@ -12,23 +12,39 @@
...
+monitoredProjects, _ := res["monitoredProjects"].([]interface{})
 
-name := res["name"].(string)
-name = tpgresource.GetResourceNameFromSelfLink(name)
+// Convert configured terraform monitored_project resource name to a ProjectNumber
+expectedProject, configProjectErr := config.NewResourceManagerClient(config.UserAgent).Projects.Get(d.Get("name").(string)).Do()
+if configProjectErr != nil {
+	return nil, configProjectErr
+}
+expectedProjectNumber := strconv.FormatInt(expectedProject.ProjectNumber, 10)
```
Comment (hao-nan-li) [inline mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: Based on the error "**Parent id must be numeric.**", I wonder something needs to be changed here.

Comment (etruong42) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: Can you elaborate? It looks like the failure came from a "google_project" resource, I don't think a decoder for the google_monitoring_monitored_project would be part of the culprit?

Comment (hao-nan-li) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: How would you approach this problem, if these tests passed previously

Comment (etruong42) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: I've never got these tests to pass locally. Other reviewers told me not to worry about it. It doesn't seem like these tests are very repeatable/reproducible in different environments. It does not seem to have the same issue in the CI tests.

Comment (hao-nan-li) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: Do you mean that even before the change these tests still fail?

Comment (etruong42) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: How do we find out what the cause is?

And yes, if I revert my changes, the test still fails.

Comment (hao-nan-li) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: https://github.com/GoogleCloudPlatform/magic-modules/pull/8454.
https://github.com/GoogleCloudPlatform/magic-modules/pull/8322. 

Looks like these tests passed before. 
The error is still most likely introduced by changes from this PR...

Comment (etruong42) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: Can we get the debug logs for what these tests currently output on the main branch?

Comment (hao-nan-li) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: https://paste.googleplex.com/6652436600061952

Comment (hao-nan-li) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: I'm looking into the delete_url of monitoringProject.yaml, see if we can properly change that to fit the right delete url.

Comment (hao-nan-li) [reply mmv1/templates/terraform/decoders/monitoring_monitored_project.go.erb:26]: The delete works after changing the delete_url to **v1/{{name}}**

File: mmv1/products/monitoring/MonitoredProject.yaml [hunk 4]
```diff
@@ -24,9 +24,6 @@ references: !ruby/object:Api::Resource::ReferenceLinks
   guides:
     'Official Documentation': 'https://cloud.google.com/monitoring/settings/manage-api'
   api: 'https://cloud.google.com/monitoring/api/ref_v3/rest/v1/locations.global.metricsScopes.projects'
```
Comment (hao-nan-li) [inline mmv1/products/monitoring/MonitoredProject.yaml:26]: Line 18 in this file. Delete url to **v1/{{name}}**

Comment (etruong42) [reply mmv1/products/monitoring/MonitoredProject.yaml:26]: Thanks! Is there an explanation for why this change is now necessary?

For posterity, here's the conversation that I think led to this recommendation:

> I think the problem comes down to whatever is decoding the deletion. In the latest CI run of `TestAccMonitoringMonitoredProject_monitoringMonitoredProjectBasicExample`, the deletion request is 

> `DELETE /v1/locations/global/metricsScopes/ci-test-project-188019/projects/locations/global/metricsScopes/1067888929963/projects/102753631080`

> and we can see the path has the format `v1/locations/global/metricsScopes/<project>/projects/<metric_scope_path>`

> when we just want `DELETE /v1/<metric_scope_path>` which would translate to `DELETE /v1/locations/global/metricsScopes/1067888929963/projects/10275363108`. I added debug lines to the decoder, and the decoder seems to decode the link just fine.

> This leads me to suspect some spooky complexity from deleting the `nested_query` functionality, in which case I think I definitely need more support in debugging

Comment (hao-nan-li) [reply mmv1/products/monitoring/MonitoredProject.yaml:26]: It's mainly due to to deleting the `nested_query`. I don't know why it was implemented with `nested_query` before. Let the CI runs complete and see if the proposed solution solved the problem for all failing tests.

**Review by hao-nan-li — APPROVED** [review]
